pkg/distributor/influx.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/mimir/pkg/distributor/influxpush"
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/util"
+	utillog "github.com/grafana/mimir/pkg/util/log"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+func parser(ctx context.Context, r *http.Request, maxSize int, _ *util.RequestBuffers, req *mimirpb.PreallocWriteRequest, logger log.Logger) error {
+	spanLogger, ctx := spanlogger.NewWithLogger(ctx, logger, "Distributor.InfluxHandler.decodeAndConvert")
+	defer spanLogger.Span.Finish()
+
+	spanLogger.SetTag("content_type", r.Header.Get("Content-Type"))
+	spanLogger.SetTag("content_encoding", r.Header.Get("Content-Encoding"))
+	spanLogger.SetTag("content_length", r.ContentLength)
+
+	ts, bytesRead, err := influxpush.ParseInfluxLineReader(ctx, r, maxSize)
+	// TODO(alexg): one argument for splitting up the decoding and conversion is to facilitate granular timings
+	// right now since ParseInfluxLineReader() does both
+	// The otel version decodes the whole input and then processes it, the existing Influx code parses each line as it
+	// decodes it.
+	level.Debug(spanLogger).Log(
+		"msg", "decodeAndConvert complete",
+		"bytesRead", bytesRead,
+	)
+	if err != nil {
+		level.Error(logger).Log("err", err.Error())
+		// TODO(alexg): need to pass on the http.StatusBadRequest
+		// http.Error(w, err.Error(), http.StatusBadRequest)
+		return err
+	}
+
+	// Sigh, a write API optimisation needs me to jump through hoops.
+	pts := make([]mimirpb.PreallocTimeseries, 0, len(ts))
+	for i := range ts {
+		pts = append(pts, mimirpb.PreallocTimeseries{
+			TimeSeries: &ts[i],
+		})
+	}
+
+	level.Debug(spanLogger).Log(
+		"msg", "Influx to Prometheus conversion complete",
+		"metric_count", len(ts),
+	)
+
+	req.Timeseries = pts
+	return nil
+}
+
+// InfluxHandler is a http.Handler which accepts Influx Line protocol and converts it to WriteRequests.
+func InfluxHandler(
+	maxRecvMsgSize int,
+	requestBufferPool util.Pool,
+	sourceIPs *middleware.SourceIPExtractor,
+	retryCfg RetryConfig,
+	push PushFunc,
+	_ *PushMetrics, // TODO(alexg) add pushMetrics()
+	_ prometheus.Registerer, // TODO(alexg): add reg
+	logger log.Logger,
+) http.Handler {
+	//TODO(alexg): mirror otel.go implementation where we do decoding here rather than in parser() func?
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := utillog.WithContext(ctx, logger)
+		if sourceIPs != nil {
+			source := sourceIPs.Get(r)
+			if source != "" {
+				logger = utillog.WithSourceIPs(source, logger)
+			}
+		}
+
+		supplier := func() (*mimirpb.WriteRequest, func(), error) {
+			rb := util.NewRequestBuffers(requestBufferPool)
+			var req mimirpb.PreallocWriteRequest
+
+			if err := parser(ctx, r, maxRecvMsgSize, rb, &req, logger); err != nil {
+				// TODO(alexg): Do we even need httpgrpc here?
+				// Check for httpgrpc error, default to client error if parsing failed
```

Comment:
I don't see any httpgrpc errors being returned by `parser`

Reply:
Oops, should have got rid of that one. Thanks. Done.

Reply:
Ah, `httpgprc` is used to smuggle both the http status code and the error message out of the `supplier()` function. I've removed the misleading comment.